drafting episode 4 (#40)
carpentries-incubator/better-research-software-r: episodes/04-code-readability.md

Author: ```suggestion
 $ git add <your_changed_files>
 $ git commit -m "Your commit message"
 $ git push origin main
```

      # pass
    } else {
      # Parse "H:MM" or "HH:MM" and convert to hours
      t <- strptime(tt, format = "%H:%M", tz = "UTC")

      if (is.na(t)[1]) {
        tt2 <- sub("^([0-9]):", "0\\1:", tt)
        t <- strptime(tt2, format = "%H:%M", tz = "UTC")
      }

      ttt <- (as.numeric(t$hour) * 3600 + as.numeric(t$min) * 60 + as.numeric(t$sec)) / (60 * 60)
      print(t)
      print(ttt)

      time <- c(time, ttt)

      if ("date" %in% names(data[[j]])) {
        date_str <- substr(data[[j]][["date"]], 1, 10)
        date <- c(date, as.POSIXct(strptime(date_str, format = "%Y-%m-%d", tz = "UTC")))
      } else {
        if (length(time) > 0) time <- time[-1]
      }
    }
  }

  j <- j + 1
}

t <- c(0)
for (i in time) {
  t <- c(t, t[length(t)] + i)
}

ord <- order(date)
date <- date[ord]
time <- time[ord]

plot_df <- data.frame(
  date = date,
  cumulative_time = t[-1]
)

p <- ggplot(plot_df, aes(x = date, y = cumulative_time)) +
  geom_point() +
  geom_line() +
  labs(
    x = "Year",
    y = "Total time spent in space to date (hours)"
  ) +
  theme_minimal()

ggsave(filename = g_file, plot = p, width = 9, height = 5, dpi = 300)
print(p)

close(data_f)

> close(data_t)

Author: upon close review, the version of the code I have uses base plotting here instead of ggplot

```

Let's make sure we commit our changes.

```bash
(venv_spacewalks) $ git add eva_data_analysis.R

> (venv_spacewalks) $ git commit -m "Move library calls to the top of the script"

Author: ```suggestion
 $ git add eva_data_analysis.R
 $ git commit -m "Move library calls to the top of the script"
```

    duration_str <- data[[j]][["duration"]]

    if (duration_str == "") {
      # pass
    } else {
      # Parse "H:MM" or "HH:MM" and convert to hours
      duration_dt <- strptime(duration_str, format = "%H:%M", tz = "UTC")

      if (is.na(duration_dt)[1]) {
        duration_str2 <- sub("^([0-9]):", "0\\1:", duration_str)
        duration_dt <- strptime(duration_str2, format = "%H:%M", tz = "UTC")
      }

      duration_hours <- (as.numeric(duration_dt$hour) * 3600 +
                         as.numeric(duration_dt$min) * 60 +
                         as.numeric(duration_dt$sec)) / (60 * 60)

      print(duration_dt)
      print(duration_hours)

      time <- c(time, duration_hours)

      if ("date" %in% names(data[[j]])) {
        date_str <- substr(data[[j]][["date"]], 1, 10)
        date <- c(date, as.POSIXct(strptime(date_str, format = "%Y-%m-%d", tz = "UTC")))
      } else {
        if (length(time) > 0) time <- time[-1]
      }
    }
  }

  j <- j + 1
}

cumulative_hours <- c(0)
for (i in time) {
  cumulative_hours <- c(cumulative_hours, cumulative_hours[length(cumulative_hours)] + i)
}

ord <- order(date)
date <- date[ord]
time <- time[ord]

> 

Reviewer: not a bad idea. Why don't we leave it as is. At the time of teaching we could ask the attendees if further changes could be implemented?

	
A minimal .lintr config file 

```
linters: lintr::linters_with_defaults(
  # Keep the signal high; tune to taste.
  line_length_linter = lintr::line_length_linter(100),

  # Common readability/style checks that work well in both .R and R chunks in .qmd/.Rmd
  object_name_linter = lintr::object_name_linter(),
  spaces_left_parentheses_linter = lintr::spaces_left_parentheses_linter(),
  spaces_inside_linter = lintr::spaces_inside_linter(),
  trailing_whitespace_linter = lintr::trailing_whitespace_linter(),

  # Catch likely mistakes without being too noisy
  unused_import_linter = lintr::unused_import_linter()
)

# Exclude rendered/output directories typical of Quarto projects
exclusions: list(
  "_site/",
  "_book/",
  "_freeze/",
  "docs/",
  "site_libs/",
  "renv/"
)
```

## Use existing packages from known developers
::::::::::::::::::::::::::::::::::::::::

## Use existing packages from known developers

Our script currently reads the data line-by-line from the JSON data file and uses custom code to manipulate the data. Variables of interest are stored in lists but there are more suitable data structures (e.g. dataframes or tibbles) to store data in our case.

> By choosing custom code over popular and well-tested libraries, we are making our code less readable and understandable and more error-prone. The main functionality of our code can be rewritten as follows using data frames from base R or tibbles from the `tidyverse` package to load and manipulate the data in data frames.  

Author: ```suggestion
By choosing custom code over popular and well-tested libraries, we are making our code less readable and understandable and more error-prone. The main functionality of our code can be rewritten as follows using data frames from base R or tibbles from the `tidyverse` package to load and manipulate the data in data frames.
We will also edit our code to use `ggplot2` for plotting so that it consistently uses the `tidyverse`.
```

> First, we need to install this dependency into our virtual environment.

Author: ```suggestion
First, we need to install the `tidyverse` as a dependency into our virtual environment.
```

```r
renv::install("tidyverse")

> #run this once to download tidyverse locally and add it to the lockfile 

Author: didn't we discover you need to run snapshot too?

```

Then we will edit the code to use `tibble`. For the sake of time in the workshop, we will give you the updated code. The code should now look like:

```r
library(tidyverse) #tidyverse "contains" ggplot2
library(jsonlite)
library(lubridate)

input_file  <- "./eva-data.json"
output_file <- "./eva-data.csv"
graph_file  <- "./cumulative_eva_graph.png"

eva_tbl <- jsonlite::fromJSON(input_file) |>
  as_tibble()

subset=['duration','date'])
eva_tbl <- eva_tbl |>
  mutate(
    eva  = as.numeric(eva),
    date = ymd_hms(date, quiet = TRUE)
  ) |>
  filter(!is.na(duration), duration != "", !is.na(date))

readr::write_csv(eva_tbl, output_file)

eva_tbl <- eva_tbl |>
  arrange(date)

eva_tbl <- eva_tbl |>
  mutate(
    duration_hours = {
      parts <- str_split(duration, ":", n = 2, simplify = TRUE)
      as.numeric(parts[, 1]) + as.numeric(parts[, 2]) / 60
    },
    cumulative_time = cumsum(duration_hours)
  )

p <- ggplot(eva_tbl, aes(x = date, y = cumulative_time)) +
  geom_point() +
  geom_line() +
  labs(
    x = "Year",
    y = "Total time spent in space to date (hours)"
  ) +
  theme_minimal()

ggsave(graph_file, plot = p, width = 9, height = 5, dpi = 300)
print(p)

```

>  $ renv::status() 

Author: ```suggestion
```

> ```{console}

Author: ```suggestion
```r
```

> renv::status() 
```

 $ git add eva_data_analysis.R renv.lock
 $ git commit -m "Refactor code and add tidyverse to lockfile"
 $ git push origin main
 $ git add eva_data_analysis.R renv.lock
 $ git commit -m "Refactor code and add tidyverse to lockfile"

>  $ git push origin main

Author: ```suggestion
 $ git add eva_data_analysis.R renv.lock
 $ git commit -m "Refactor code and add tidyverse to lockfile"
 $ git push origin main
```

```

We have committed the code and the environment changes together since they are related and form one logical unit of change.

## Use comments to explain functionality

Commenting is a very useful practice to help convey the context of the code. It can be helpful as a reminder for your future self or your collaborators as to why code is written in a certain way, how it is achieving a specific task, or the real-world implications of your code.

> R’s comment syntax is simple and familiar if you’ve used Python or shell scripting: a # starts a comment that runs to the end of the line, and there’s no distinct multi-line comment token in the base language—multi-line commentary is just multiple # lines. That contrasts with languages like C, C++, Java, and JavaScript, which typically have both // for single-line comments and /* … */ for block comments, and with SQL, which often supports -- and /* … */. Python is closest to R in practice (# for line comments), but Python also has the convention of triple-quoted strings used as docstrings, which can function like structured, extractable documentation; R’s comparable “docstring” mechanism is usually done via `roxygen2` comments (#' ...) in packages rather than a core language feature.

Author: ```suggestion
```
a lot on other coding languages here.  Think we should keep to R and not assume people know other languages

From the official [R documentation](https://cran.r-project.org/doc/manuals/r-patched/R-lang.html?utm_source=chatgpt.com#Comments-1): 

> Comments in R are ignored by the parser. Any text from a # character to the end of the line is taken to be a comment, unless the # character is inside a quoted string. For example,

> x <- 1  # This is a comment...
> y <- "  #... but this is not."

The [Tidyverse Style] Guide(https://style.tidyverse.org/functions.html#comments): 

> In code, use comments to explain the “why” not the “what” or “how”. Each line of a comment should begin with the comment symbol and a single space: #.

```r
x <- 5  # In R, "inline comments"" begin with the `#` symbol and at least one space

# this is a single-line comment
y <- x + 10
z <- y*2 + x

> # This is a multiline comment in Python.

Author: ```suggestion
# This is a multiline comment in R.
```

::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::: callout
How to install lintr and styler for RStudio

1) Install the packages

> 

Reviewer: I don't feel strongly about this. In fact, I don't use lintr at all when I am coding

> 

Author: Let's remove it then.

RStudio Addins

styler provides RStudio addins to:
	-	style the active file
	-	style the current package
	-	style the highlighted selection  

Find them at:
	-	Tools → Addins → Browse Addins… → search “styler”

A practical “RStudio workflow”
	1.	Write code
	2.	Run styler addin (format)
	3.	Run lintr addin (lint) and fix anything meaningful
	
A minimal .lintr config file 

```
linters: lintr::linters_with_defaults(
  # Keep the signal high; tune to taste.
  line_length_linter = lintr::line_length_linter(100),

  # Common readability/style checks that work well in both .R and R chunks in .qmd/.Rmd
  object_name_linter = lintr::object_name_linter(),
  spaces_left_parentheses_linter = lintr::spaces_left_parentheses_linter(),
  spaces_inside_linter = lintr::spaces_inside_linter(),
  trailing_whitespace_linter = lintr::trailing_whitespace_linter(),

  # Catch likely mistakes without being too noisy
  unused_import_linter = lintr::unused_import_linter()
)

# Exclude rendered/output directories typical of Quarto projects
exclusions: list(
  "_site/",
  "_book/",
  "_freeze/",
  "docs/",
  "site_libs/",
  "renv/"
)
```

::::::::::::::::::::::::::::::::::::::::

> ## Use packages 

Author: ```suggestion
## Use existing packages from known developers
```

> Our script currently reads the data line-by-line from the JSON data file and uses custom code to manipulate the data. Variables of interest are stored in lists but there are more suitable data structures (e.g. dataframes or tibbles) to store data in our case.

Author: I think this section might actually be better converting the base R plotting to use ggplot?  I think we installed it in the last episode so we could do this later?

Reviewer: Maybe? I defer to you on this

output_file <- "./eva-data.csv"
graph_file  <- "./cumulative_eva_graph.png"

eva_tbl <- jsonlite::fromJSON(input_file) |>
  as_tibble()

subset=['duration','date'])
eva_tbl <- eva_tbl |>
  mutate(
    eva  = as.numeric(eva),
    date = ymd_hms(date, quiet = TRUE)
  ) |>
  filter(!is.na(duration), duration != "", !is.na(date))

readr::write_csv(eva_tbl, output_file)

eva_tbl <- eva_tbl |>
  arrange(date)

eva_tbl <- eva_tbl |>
  mutate(
    duration_hours = {
      parts <- str_split(duration, ":", n = 2, simplify = TRUE)
      as.numeric(parts[, 1]) + as.numeric(parts[, 2]) / 60
    },
    cumulative_time = cumsum(duration_hours)
  )

p <- ggplot(eva_tbl, aes(x = date, y = cumulative_time)) +
  geom_point() +
  geom_line() +
  labs(
    x = "Year",
    y = "Total time spent in space to date (hours)"
  ) +
  theme_minimal()

ggsave(graph_file, plot = p, width = 9, height = 5, dpi = 300)
print(p)

```

Once we have replaced the  code in our Python script `eva_data_analysis.R` with the above code. As we installed the packages with renv::install() function, the changes are captured in the lockfile. We can verify that it is the case by running on the console 

```{console}

> (venv_spacewalks) $ renv::status() 

Author: ### 
```suggestion
> renv::status() 
```

Some good comments may look like the example below.

```r

library(tidyverse)
library(jsonlite)
library(lubridate)

# Files
input_file  <- "./eva-data.json"
output_file <- "./eva-data.csv"
graph_file  <- "./cumulative_eva_graph.png"

# 1) Read JSON array into a tibble
eva_tbl <- jsonlite::fromJSON(input_file) |>
  as_tibble()

> # 2) Convert types + drop missing duration/date (pandas dropna subset=['duration','date'])

Author: ```suggestion
# 2) Convert types + drop missing duration/date
```

eva_tbl <- eva_tbl |>
  mutate(
    eva  = as.numeric(eva),
    date = ymd_hms(date, quiet = TRUE)
  ) |>
  filter(!is.na(duration), duration != "", !is.na(date))

# 3) Write CSV (index=False equivalent)
readr::write_csv(eva_tbl, output_file)

# 4) Sort by date
eva_tbl <- eva_tbl |>
  arrange(date)

# 5) duration_hours + cumulative_time
eva_tbl <- eva_tbl |>
  mutate(
    duration_hours = {
      parts <- str_split(duration, ":", n = 2, simplify = TRUE)
      as.numeric(parts[, 1]) + as.numeric(parts[, 2]) / 60
    },
    cumulative_time = cumsum(duration_hours)
  )

# 6) Plot + save
p <- ggplot(eva_tbl, aes(x = date, y = cumulative_time)) +
  geom_point() +
  geom_line() +
  labs(
    x = "Year",
    y = "Total time spent in space to date (hours)"
  ) +
  theme_minimal()

ggsave(graph_file, plot = p, width = 9, height = 5, dpi = 300)
print(p)

```

> Note that we have also added some useful print statements, to let us know what stage the analysis is in.

Author: ```suggestion
```
we are no longer doing logging with print statements in the script above

Commit changes:

```bash
(venv_spacewalks) $ git add eva_data_analysis.R
(venv_spacewalks) $ git commit -m "Add comments to the code"

> (venv_spacewalks) $ git push origin main

Author: ```suggestion
 $ git add eva_data_analysis.R
 $ git commit -m "Add comments to the code"
 $ git push origin main
```

## Separate units of functionality

Functions are a fundamental concept in writing software and are one of the core ways you can organize your code to improve its readability. A function is an isolated section of code that performs a single, *specific* task that can be simple or complex.

A function can then be called multiple times with different inputs throughout a codebase, but its definition only needs to 
appear once. Breaking up code into functions in this manner benefits readability since the smaller sections are easier to read 
and understand.

> Because functions are reusable, they naturally encourage the [Don’t Repeat Yourself (DRY) principle](https://adv-r.had.co.nz/Functional-programming.html?utm_source=chatgpt.com): instead of copying the same logic throughout a codebase, you define it once and call it as needed. This keeps software from becoming unnecessarily long and confusing. It also improves maintainability—when the behavior in a function needs to change, you update it in one place rather than chasing down multiple copies.

Author: ```suggestion
Because functions are reusable, they naturally encourage the [Don’t Repeat Yourself (DRY) principle](https://glosario.carpentries.org/en/#dry): instead of copying the same logic throughout a codebase, you define it once and call it as needed. This keeps software from becoming unnecessarily long and confusing. It also improves maintainability—when the behavior in a function needs to change, you update it in one place rather than chasing down multiple copies.
```

The link didn't work.  There is a section on functional prgramming in advanced R but this is the link  - https://adv-r.hadley.nz/fp.html - and it doesn't mention DRY.

  jsonlite::fromJSON(input_file) |>
    tibble::as_tibble()
}

# 2) Convert + write to CSV (returns the cleaned df invisibly for chaining)
write_dataframe_to_csv <- function(df, output_file) {
  df <- df |>
    dplyr::mutate(
      eva  = as.numeric(eva),
      date = lubridate::ymd_hms(date, quiet = TRUE)
    ) |>
    dplyr::filter(!is.na(duration), duration != "", !is.na(date))

  readr::write_csv(df, output_file)
  df
}

# 3) Plot cumulative time in space and save the figure
plot_cumulative_time_in_space <- function(df, graph_file) {
  df <- df |>
    dplyr::arrange(date) |>
    dplyr::mutate(
      duration_hours = {
        parts <- stringr::str_split(duration, ":", n = 2, simplify = TRUE)
        as.numeric(parts[, 1]) + as.numeric(parts[, 2]) / 60
      },
      cumulative_time = cumsum(duration_hours)
    )

  p <- ggplot2::ggplot(df, ggplot2::aes(x = date, y = cumulative_time)) +
    ggplot2::geom_point() +
    ggplot2::geom_line() +
    ggplot2::labs(
      x = "Year",
      y = "Total time spent in space to date (hours)"
    ) +
    ggplot2::theme_minimal()

  ggplot2::ggsave(graph_file, plot = p, width = 9, height = 5, dpi = 300)

>   print(p)

Author: It looks like both the data cleaning and the plotting are in the same function.  Maybe we should add them as two functions?

  invisible(p)
}

# --- Main (now simplified) ---
eva_tbl <- read_json_to_dataframe(input_file) |>
  write_dataframe_to_csv(output_file = output_file)

plot_cumulative_time_in_space(eva_tbl, graph_file)

```
:::

:::

## Use `roxygen2` comments to document functions

Now that we’ve written a few functions, it’s time to document them so we can quickly remember what they do. That way, someone reading this code later can understand the intent without having to reverse-engineer the implementation.

> In R, there isn’t a built-in “docstring” feature in the same way Python has. In scripts, the usual approach is to write clear comments directly above a function. In packages, the standard is to use [`roxygen2`](https://`roxygen2`.r-lib.org/): a structured comment block (lines starting with #') that describes what the function does, what inputs it expects, what it returns, and any important edge cases.

Author: ```suggestion
In scripts, the usual approach is to write clear comments directly above a function. In packages, the standard is to use [`roxygen2`](https://`roxygen2`.r-lib.org/): a structured comment block (lines starting with #') that describes what the function does, what inputs it expects, what it returns, and any important edge cases.
```

Good function documentation improves readability by making the purpose of the code explicit. It also makes functions easier to reuse: when the documentation clearly spells out inputs and outputs, you can call the function elsewhere with confidence—without re-reading the whole body to figure out what it needs and what it produces.

> With `roxygen2`, the documentation is written immediately above the function definition using #' comments, and tooling can turn that into the help text you see via ?function_name when you’re working in a package. That said, you can still use roxygen-style comments even if you’re not building a formal package. In a regular project, this looks like adopting #' blocks as a consistent “docstring convention” in your .R scripts (or in a dedicated R/ folder), using tags like @param, @return, and @examples to standardize what you record. You won’t automatically get ?my_function help pages unless you generate Rd files as part of a package build, but you still gain a structured, readable, and machine-friendly format that’s easy to search and maintain—and if the project ever does become a package, most of the documentation work is already done.

Author: ```suggestion
With `roxygen2`, the documentation is written immediately above the function definition using #' comments, and tooling can turn that into the help text you see via ?function_name when you’re working in a package. That said, you can still use roxygen-style comments even if you’re not building a formal package. In a regular project, this looks like adopting #' blocks as a consistent convention in your .R scripts (or in a dedicated R/ folder), using tags like @param, @return, and @examples to standardize what you record. You won’t automatically get ?my_function help pages unless you generate Rd files as part of a package build, but you still gain a structured, readable, and machine-friendly format that’s easy to search and maintain—and if the project ever does become a package, most of the documentation work is already done. We will learn to make the Rd files later but go ahead and start adding comments now.
```

what does "using tags like @param, @return, and @examples to standardize what you record" mean?

Author: we will need to write more about what the tags do...

Might be worth doing minimal here and then discussing more in the documentation episode later.

> ### Example of a single-line docstring

Author: ```suggestion
### Example of a single-line comment
```

```r

add <- function(x, y) {
  # Add two numbers together
  x + y
}

```

> ### Example of a multi-line docstring

Author: ```suggestion
### Example of a multi-line `roxygen2` comment
```

```r

#' Add two numbers together
#'
#' @param x A numeric value.
#' @param y A numeric value.
#' @return A single numeric value equal to `x + y`.
#' @examples
#' add(1, 2)
#' add(3.5, 4)
add <- function(x, y) {
  x + y
}

```

Some projects may have their own guidelines on how to write `roxygen2` comments, such as [The Tidyverse Style Guide ](https://style.tidyverse.org/documentation.html). If you are contributing code to a wider project or community, try to follow the guidelines and standards they provide 
for code style.

As your code grows and becomes more complex, the `roxygen2` comments can form the content of a reference guide allowing developers to quickly look up how to use the APIs, functions, and classes defined in your codebase.

> In R, it’s common to use tooling that extracts function documentation from your source and publishes it as a browsable website, so people can learn how to use your code without cloning the repo or reading the raw files. For package-style documentation written with `roxygen2`, a typical workflow is to generate reference pages from your #' comments and build a site with pkgdown, which produces a documentation website from your package and its help files.

Author: ```suggestion
In R, it’s common to use tooling that extracts function documentation from your source and publishes it as a browsable website, so people can learn how to use your code without cloning the repo or reading the raw files. For package-style documentation written with `roxygen2`, a typical workflow is to generate reference pages from your #' comments and build a site with pkgdown, which produces a documentation website from your package and its help files. We will do this in a later episode that focuses on documentation.
```

Let's write `roxygen2` comments for the function `read_json_to_dataframe` we introduced in the previous exercise. Remember, questions we want to answer when writing the `roxygen2` comments include:

- What the function does?
- What kind of inputs does the function take? Are they required or optional? Do they have default values?
- What output will the function produce?
- What exceptions/errors, if any, it can produce?

Our `read_json_to_dataframe` function fully described by the `roxygen2` comments may look like:

```r

#' Read EVA data from a JSON file into a tibble
#'
#' Reads a JSON file containing an array of records (objects) and returns the
#' contents as a tibble for downstream analysis.
#'
#' @param input_file Path to a JSON file (character scalar). The file is expected
#'   to contain a JSON array of objects, e.g. `[{"eva":"1", ...}, {"eva":"2", ...}]`.
#' @return A tibble with one row per JSON record and one column per field.
#' @examples
#' \dontrun{
#' eva_tbl <- read_json_to_dataframe("./eva-data.json")
#' dplyr::glimpse(eva_tbl)

> #' }

Author: might need to explain this `\dontrun{}`1 syntax somewhere before we use it

## Summary

Good code readability brings many benefits to software development. It makes code easier to understand, maintain, and debug—helping collaborators and future developers as well as the original author. Readable code reduces the risk of errors, speeds up onboarding for new team members, and simplifies code reviews. It also supports long-term sustainability: clear code is easier to extend, adapt, and refactor over time.

Integrated development environments (IDEs) can further improve readability by surfacing issues early through built-in or integrated static analysis tools (linters). Linters flag common problems and style violations as you write code, which helps you address issues immediately rather than discovering them later in the development cycle.

> Case in point, in RStudio, you can add `roxygen2` documentation blocks quickly using built-in helpers. After installing and loading `roxygen2`, place your cursor inside (or immediately above) a function and use Code → Insert Roxygen Skeleton to generate a template comment block with #' lines and common tags like @param and @return. RStudio will pre-populate the parameter list based on the function signature, so you can focus on filling in descriptions and examples. Many people also bind a keyboard shortcut to “Insert Roxygen Skeleton” so documenting functions becomes part of the normal edit cycle, and then run Document (commonly via devtools) to regenerate the help files when working in a package.

Author: I don't think we talked about roxygen skeletons above?  maybe we should mention instead of only including in summary?
